<a href="https://colab.research.google.com/github/MarkDvoryanchikov/Numerical_analysis_2020/blob/master/3_%D0%A0%D0%B5%D1%88%D0%B5%D0%BD%D0%B8%D0%B5_%D1%81%D0%B8%D1%81%D1%82%D0%B5%D0%BC_%D0%BB%D0%B8%D0%BD%D0%B5%D0%B9%D0%BD%D1%8B%D1%85_%D0%B0%D0%BB%D0%B3%D0%B5%D0%B1%D1%80%D0%B0%D0%B8%D1%87%D0%B5%D1%81%D0%BA%D0%B8%D1%85_%D1%83%D1%80%D0%B0%D0%B2%D0%BD%D0%B5%D0%BD%D0%B8%D0%B9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Решение систем линейных алгебраических уравнений**

## **Векторы, матрицы и базовые операции над ними**

### **Векторы**

Python списки по умолчанию не являются векторами. Для обращения с ними как с векторами необходимо определять специальные функции или обходить их поэлементно.

In [ ]:
list_1 = [10, 20, 20]
list_2 = [10, 10, 10]

In [ ]:
sum_1_2 = list_1 + list_2
sum_1_2

[10, 20, 20, 10, 1, 10]

Происходит конкатенация списков, а не их векторное сложение.
Чтобы сложить списки как векторы, можно написать и использовать функцию:

In [ ]:
def add_vectors(v1, v2):          # на вход функции падаётся 2 объекта вектора
    return [v1i + v2i for v1i, v2i in zip(v1, v2)] # [10 + 10, 20 + 10, 20 + 10]
    # zip - функция сопоставляющая элементы последовательностей/списков друг с другом. 
    # n-й элемент v1 соответсвует n-му v2, что позволяет компактно поэлементено обходить несколько списко сразу.

add_vectors(list_1, list_2)

[20, 30, 30]

Эти же действия можно сделать с использованием **numpy** несколько проще:

In [ ]:
import numpy as np

In [ ]:
np_list_1 = np.array([10, 20, 20])
np_list_2 = np.array([10, 10, 10])

np_list_1 + np_list_2 # [10 + 10, 20 + 10, 20 + 10]

array([20, 30, 30])

**Попробуйте выполнить другие арифметические операции с векторами самостоятельно:**


---



**А что будет если попробовать выполнить такие операции к векторам разной длины?**

---



#### **Умножение вектора на число**
Попробуем применить к списку python:



In [ ]:
list_1 = [10, 20, 20]
list_1 * 2

[10, 20, 20, 10, 20, 20]

Список стал в 2 раза больше, но при умножении вектора на число происходит его поэлементное умножение. Этого можно достичь обойдя список в цикле.

In [ ]:
list_1 = [10, 20, 20]
list_mult = [i * 2 for i in list_1]
list_mult

[20, 40, 40]

Вернёмся к numpy:

In [ ]:
np_list_1 = np.array([10, 20, 20])
np_list_1 * 2

array([20, 40, 40])

Несколько полезных функций для векторов:

In [ ]:
np_list_1 = np.array([10, 20, 20])
print("Сумма элементов:", sum(np_list_1))
print("Минимальный элемент:", min(np_list_1))
print("Максимальный элемент:", max(np_list_1))

Сумма элементов: 50
Минимальный элемент: 10
Максимальный элемент: 20


### **Матрицы**

Матрицу в numpy можно определить следующим образом:

In [ ]:
matrix = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    ])
matrix

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9]])

Попробуем сложить 2 матрицы:

In [ ]:
matrix1 = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    ])
matrix2 = np.array([
    [1, 2, 3],
    [4, 5, 6],
    ])
matrix1 + matrix2

ValueError: ignored

**Всё верно!** Складывать и вычитать можно только матрицы одинакового размера:

In [ ]:
matrix1 = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    ])
matrix2 = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    ])
matrix1 + matrix2

array([[ 2,  4,  6],
       [ 8, 10, 12],
       [14, 16, 18]])

Сложение матрицы и вектора:

In [ ]:
matrix1 = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    ])
vector1 = np.array([1, 2, 3])
matrix1 + vector1

array([[ 2,  4,  6],
       [ 5,  7,  9],
       [ 8, 10, 12]])

**Поэксперементируйте с арифметическим операциями над матрицами самостоятельно:**


---



Поработаем с конкретной (первой) строкой матрицы:

In [ ]:
matrix1 = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    ])
vector1 = np.array([1, 2, 3])
matrix1[0] = matrix1[1] + vector1*10
matrix1

array([[14, 25, 36],
       [ 4,  5,  6],
       [ 7,  8,  9]])

**Используя базовые операции привести систему к верхней треугольной матрице (метод Гаусса - Прямой ход):**


---


In [ ]:
a = np.array([
    [2, 1, -1],
    [-3, -1, 2],
    [-2, 1, 2],
    ])
b = np.array([8, -11, -3])

**Используя обратный ход найти корни полученной системы на предыдущем шаге (вывести вектор найденных неизвестных)**


---



### **Нормы матрицы**

In [ ]:
A = np.array([
    [1, 2, -1],
    [-3, -3, 4],
    [-5, 2, 2],
    ])

Для матрицы **A** найти норму **$ ||A||_1 $**

---



Для матрицы **A** найти норму **$ ||A||_2 $**

Для матрицы **A** найти норму **$ ||A||_{\infty} $**

In [ ]:
# Подсказка. Получить первый столбец матрицы можно следующим образом: С[:,0]

## **Индивидуальное задание**

Необходимо решить систему уравнений записанную в матричной форме (1) **методом простых итераций** и (2) **методом Зейделя**. Представить полученные значения для каждой итерации. Сравнить полученное количество итераций. Какой метод и насколько получился быстрее?

Точность вычислений: **ɛ < 0.001**

### **Системы:**

1) $A = \begin{bmatrix}
       8 & 2 & 1           \\[0.7em]
       3 & 7 & -1          \\[0.7em]
       2 & 1 & -9          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       3           \\[0.7em]
       4           \\[0.7em]
       6           \\[0.7em]      
     \end{bmatrix}$; 

2) $A = \begin{bmatrix}
       7 & 2 & 1           \\[0.7em]
       4 & 10 & -1          \\[0.7em]
       2 & 1 & -8          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       4           \\[0.7em]
       5           \\[0.7em]
       6           \\[0.7em]      
     \end{bmatrix}$; 

3) $A = \begin{bmatrix}
       7 & 2 & -1           \\[0.7em]
       3 & 8 & 1          \\[0.7em]
       -2 & 2 & -9          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       4           \\[0.7em]
       2           \\[0.7em]
       6           \\[0.7em]      
     \end{bmatrix}$ 
     
4) $A = \begin{bmatrix}
       7 & 2 & -2           \\[0.7em]
       2 & 8 & 3          \\[0.7em]
       2.5 & -2 & 4          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       3.6           \\[0.7em]
       4.2           \\[0.7em]
       -2            \\[0.7em]      
     \end{bmatrix}$

5) $A = \begin{bmatrix}
       8 & 4 & 3           \\[0.7em]
       3 & 10 & 4          \\[0.7em]
       4 & 5 & 11          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       5           \\[0.7em]
       3           \\[0.7em]
       9            \\[0.7em]      
     \end{bmatrix}$ 
     
6) $A = \begin{bmatrix}
       7 & 2.5 & 3           \\[0.7em]
       2 & 9 & -5          \\[0.7em]
       4 & 2 & 10          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       7           \\[0.7em]
       -6           \\[0.7em]
       6            \\[0.7em]      
     \end{bmatrix}$ 
     
7) $A = \begin{bmatrix}
       6 & 2.5 & 3           \\[0.7em]
       1 & 9 & -5          \\[0.7em]
       5 & 2 & 11          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       3           \\[0.7em]
       -2           \\[0.7em]
       5            \\[0.7em]      
     \end{bmatrix}$ 
     
8) $A = \begin{bmatrix}
       8 & 3 & -2           \\[0.7em]
       2.5 & 7 & 3          \\[0.7em]
       2.5 & -2 & 9          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       3.6           \\[0.7em]
       4.4           \\[0.7em]
       -2            \\[0.7em]      
     \end{bmatrix}$

9) $A = \begin{bmatrix}
       9 & 2.5 & 3           \\[0.7em]
       -1 & 10 & -5          \\[0.7em]
       5 & 5 & 12          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       2           \\[0.7em]
       -4           \\[0.7em]
       12            \\[0.7em]      
     \end{bmatrix}$

10) $A = \begin{bmatrix}
       -8 & 2.5 & -3           \\[0.7em]
       1.5 & 9 & -5          \\[0.7em]
       5 & 5 & 11          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       4           \\[0.7em]
       3           \\[0.7em]
       -8            \\[0.7em]      
     \end{bmatrix}$

11) $A = \begin{bmatrix}
       -9 & 5 & -3           \\[0.7em]
       2.5 & 12.2 & -5          \\[0.7em]
       1.5 & 5 & 10          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       -4.3           \\[0.7em]
       3           \\[0.7em]
       6            \\[0.7em]      
     \end{bmatrix}$

12) $A = \begin{bmatrix}
       -9 & 5 & -3           \\[0.7em]
       2 & 8.5 & -5          \\[0.7em]
       1.5 & 5 & 9          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       3           \\[0.7em]
       -2           \\[0.7em]
       -5            \\[0.7em]      
     \end{bmatrix}$

13) $A = \begin{bmatrix}
       8 & 2.7 & 4           \\[0.7em]
       3.5 & 7 & 3          \\[0.7em]
       1.5 & 4 & -9          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       7.8           \\[0.7em]
       -2           \\[0.7em]
       -4            \\[0.7em]      
     \end{bmatrix}$

14) $A = \begin{bmatrix}
       -10 & 5 & -3           \\[0.7em]
       3.5 & 10.5 & -5          \\[0.7em]
       1.5 & 5 & 8          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       -3           \\[0.7em]
       2           \\[0.7em]
       5            \\[0.7em]      
     \end{bmatrix}$

15) $A = \begin{bmatrix}
       9 & 2.5 & 3           \\[0.7em]
       -1 & 10 & -5          \\[0.7em]
       5 & 5 & 12          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       2           \\[0.7em]
       -3           \\[0.7em]
       18            \\[0.7em]      
     \end{bmatrix}$

16) $A = \begin{bmatrix}
       8 & 2.5 & 3           \\[0.7em]
       3 & 10 & -4          \\[0.7em]
       4 & 2 & 10          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       5           \\[0.7em]
       -3           \\[0.7em]
       9            \\[0.7em]      
     \end{bmatrix}$

17) $A = \begin{bmatrix}
       -13 & 6 & -3           \\[0.7em]
       3 & 12.5 & -4          \\[0.7em]
       1.5 & 4 & 9          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       4           \\[0.7em]
       -3           \\[0.7em]
       -5            \\[0.7em]      
     \end{bmatrix}$

18) $A = \begin{bmatrix}
       -9 & 5 & -3           \\[0.7em]
       2 & 13.5 & -4          \\[0.7em]
       1.5 & 3 & 8          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       -4           \\[0.7em]
       3           \\[0.7em]
       -5            \\[0.7em]      
     \end{bmatrix}$

19) $A = \begin{bmatrix}
       -8 & 2.5 & -3           \\[0.7em]
       1.5 & 9 & -5          \\[0.7em]
       5 & 5 & 12          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       -4           \\[0.7em]
       3           \\[0.7em]
       8            \\[0.7em]      
     \end{bmatrix}$

20) $A = \begin{bmatrix}
       -7 & 5 & -3           \\[0.7em]
       1.5 & 12.2 & -5          \\[0.7em]
       1.5 & 5 & 12          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       -4.3           \\[0.7em]
       3.3           \\[0.7em]
       6            \\[0.7em]      
     \end{bmatrix}$

21) $A = \begin{bmatrix}
       -11 & 2.5 & -3           \\[0.7em]
       -1 & 8 & -5          \\[0.7em]
       5 & 5 & 12          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       4           \\[0.7em]
       -3           \\[0.7em]
       8            \\[0.7em]      
     \end{bmatrix}$

22) $A = \begin{bmatrix}
       3 & 2.5 & 3           \\[0.7em]
       2 & 8 & -5          \\[0.7em]
       4 & 2 & 10          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       3           \\[0.7em]
       -2           \\[0.7em]
       5            \\[0.7em]      
     \end{bmatrix}$

23) $A = \begin{bmatrix}
       9 & 4 & 3           \\[0.7em]
       3 & 10 & -4          \\[0.7em]
       4 & 2 & 8          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       5           \\[0.7em]
       -3           \\[0.7em]
       9            \\[0.7em]      
     \end{bmatrix}$

24) $A = \begin{bmatrix}
       7 & 2.5 & 3           \\[0.7em]
       2 & 8 & -5          \\[0.7em]
       4 & 2 & 11          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       5           \\[0.7em]
       -5           \\[0.7em]
       6            \\[0.7em]      
     \end{bmatrix}$

25) $A = \begin{bmatrix}
       8 & 2.5 & 3           \\[0.7em]
       2 & 10 & -4          \\[0.7em]
       4 & 2 & 7          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       5           \\[0.7em]
       -3           \\[0.7em]
       9            \\[0.7em]      
     \end{bmatrix}$

## **Разбор примера (Вариант 25)**

$A = \begin{bmatrix}
       8 & 2.5 & 3           \\[0.7em]
       2 & 10 & -4          \\[0.7em]
       4 & 2 & 7          \\[0.7em]
     \end{bmatrix} b = \begin{bmatrix}
       5           \\[0.7em]
       -3           \\[0.7em]
       9            \\[0.7em]      
     \end{bmatrix}$ **ɛ < 0.001**

Запишем в виде системы уравнений:

$$ \begin{cases} 
                    8x_1 + 2.5x_2 + 3x_3 = 5 \\[0.7em]
                    2x_1 + 10x_2  - 4x_3 = -3 \\[0.7em]
                    4x_1 + 2x_2 + 7x_3 = 9 \\[0.7em]
\end{cases} $$

Теперь преобразуем к удобному для итераций виду **$x = Cx + f$**:

$$ \begin{cases} 
                    x_1 = \frac{1}{8}(5 - 2.5x_2 - 3x_3) \\[0.7em]
                    x_2 = \frac{1}{10}(-3 - 2x_1 + 4x_3) \\[0.7em]
                    x_3 = \frac{1}{7}(9 - 4x_1 - 2x_2) \\[0.7em]
\end{cases} $$

In [ ]:
print(1/8*(5), 1/8*(-2.5), 1/8*(3),"\n",
      1/10*(-3), 1/10*(-2), 1/10*(4),"\n",
      1/7*(9) ,1/7*(-4), 1/7*(-2))

0.625 -0.3125 0.375 
 -0.30000000000000004 -0.2 0.4 
 1.2857142857142856 -0.5714285714285714 -0.2857142857142857


$$ \begin{cases} 
                    x_1 = 0.625 - 0.3125x_2 - 0.375x_3 \\[0.7em]
                    x_2 = -0.3 - 0.2x_1 + 0.4x_3 \\[0.7em]
                    x_3 = 1.28571 - 0.57143x_1 - 0.28571x_2 \\[0.7em]
\end{cases} $$

Метод итерации применяют в случае, если сходится последовательность приближений по указанному алгоритму - норма матрицы принимает значение < 1



$$C = \begin{bmatrix}
       0 & -0.3125 & -0.375           \\[0.7em]
       -0.2 & 0 & 0.4                 \\[0.7em]
       -0.57143 & -0.28571 & 0        \\[0.7em]
     \end{bmatrix}$$

Вычислим **$ ||C||_1 $**

In [ ]:
C = np.array([
              [0, -0.3125, 0.375],
              [-0.2, 0, 0.4],
              [-0.57143, -0.28571, 0],
])
C

array([[ 0.     , -0.3125 ,  0.375  ],
       [-0.2    ,  0.     ,  0.4    ],
       [-0.57143, -0.28571,  0.     ]])

In [ ]:
normC = max(sum(abs(C[0])),sum(abs(C[1])),sum(abs(C[2])))
normC

0.85714

**$ ||C||_1 $ < 1**  - следует метод должен сойтись.

Запишем уравнения и найдём первое приближение подставив в правую часть вместо $x_1, x_2, x_3$ вектор свободных членов $f$


In [ ]:
x1_0 = 0.625
x2_0 = -0.3
x3_0 = -1.28571

x1_1 = 0.625 - 0.3125*x2_0 - 0.375*x3_0
x2_1 = -0.3 - 0.2*x1_0 + 0.4*x3_0
x3_1 = 1.28571 - 0.57143*x1_0 - 0.28571*x2_0

print(x1_1, x2_1, x3_1)

1.20089125 -0.939284 1.01427925


Оценим погрешность полученную в результате первого приближения
$\frac{q}{1-q}||x^{(k)} - x^{(k-1)}||$, где $q = ||C||$

In [ ]:
err = normC/(1-normC)*max(abs(x1_0-x1_1),abs(x2_0-x2_1),abs(x3_0-x3_1))
err

13.79961350794484

***13.79961 > 0.001*** Перейдём ко второму и следующим приближениям, пока не будет достигнута требуемая точность.

$\color{red}{\text{Чтобы не было совсем скучно, допишите код вычислений самостоятельно. Используйте циклы.}}$


**Метод Зейделя**

Метод является модификацией метода простой итерации. Единственное
различие в том, что при вычислении (k + 1)-го приближения неизвестного
𝒙𝒊 при i > 1 используются уже ранее вычисленные к этому моменту (k + 1)
приближения неизвестных на текущей итерации:

$$ \begin{cases} 
                    x_1^{(k+1)} = 0.625 - 0.3125x_2^{(k)} - 0.375x_3^{(k)} \\[0.7em]
                    x_2^{(k+1)} = -0.3 - 0.2x_1^{(k+1)} + 0.4x_3^{(k)} \\[0.7em]
                    x_3^{(k+1)} = 1.28571 - 0.57143x_1^{(k+1)} - 0.28571x_2^{(k+1)} \\[0.7em]
\end{cases} $$

**Не забываем в ответе указать какой метод получился быстрее и насколько.**